Define agent architecture (#3)
dev-sys-do/sealci: docs/arch/agent.md

# Summary

The agent is the powerhouse of SealCi. It receives actions and runs them in order to complete the operational part of the CI.

# References

## Actions

An action is defined as list of commands. Actions are launched in a controlled environment, the action execution environment, as defined by the pipeline definition.

## Artifacts

During their execution, actions generate artifacts, which include:

-  Execution logs
-  Files generated during the action execution

# Agent operations

## Life of an agent

**Launching & Registering with the Scheduler**

Initially, the agent registers with a scheduler. As part of the registration process, the agent and the scheduler establish a streaming, bi-directional connection.

After the schedulers acknowledges the registration, the agent is ready to accept and process new actions.

**Health and Death**  
An agent streams health and status information to the scheduler, and the agent is kept on the scheduler's resource pool as long as it maintains its connection with it.

**Launching actions**  
Each time a action is received the agent will:

-  Create and run a container, based on the action execution environment configuration.

> -  The list of command described in the action configuration are executed in the action container.
>    -  For each command, an exit code will be returned to the scheduler. If one command fails,the next ones aren't executed and the action will be marked as failed.

Reviewer: ```suggestion
-  Execute the list of command described in the action configuration, from the action container.
   -  For each command, an exit code will be returned to the scheduler. If one command fails, the next ones aren't executed and the action will be marked as failed.
```

Reviewer 2: done

> -  Once all the action commands are completed, the agent cleans the action execution environment up by deleting its container.

Reviewer: ```suggestion
- Clean the action execution environment up by deleting its container, once all the action commands are completed.
```

Reviewer 2: done

## Action execution environment

> Each time an action execution environment will launch a session will be created that will store the state of the action. When action stop (fail/succeed) all detail will be launch to the session that will gather datas that will be sent to the scheduler.  

Reviewer: I suppose this is what you mean:

```suggestion
An action execution environment defines the context in which actions are executed, like for example a Linux container. 

 Once an action completes, its status and logs are gathered from its execution environment and forwarded to the scheduler.  
```

Reviewer 2: done

> Multiple environment can be launch in the same time to execute actions in parallel.  

Reviewer: ```suggestion
Multiple execution environments can be launched simultaneously, in order to run actions in parallel.
```

Reviewer 2: done

> Once a action is done the environment must be killed and any remains of the execution must be cleaned.

Reviewer: This is already described in the previous section, I would remove it.

# Summary
The agent is the powerhouse of SealCi. It will receive jobs and try to run them in order to complete the operational part of the CI.

# References

## Jobs
Jobs are process that are launched in a controlled environment provided from a configuration. Those process are list of choosen commands.

## Artifacts
Artifacts is the result of the execution of a job. We consider as artifacts:
- logs of execution of the pod
- any file that is created during the execution

# Agent operations

## Life of an agent

**Launching & registering to the scheduler**
At its start the agent will try to register to the scheduler in order to accept new jobs. For the acknowledgement, the agent will init a connection that will allow both the agent the scheduler to stream datas to each other.   

**Health and death**

> During the life of the agent, it will stream regularly health information of the agent. The scheduler will know that an agent is dead once the stream between them is dead.

Reviewer: There is usually no regularity in how the agent updates the scheduler with it status. The agent must only do so when its status changes.

```suggestion
An agent streams health and status information to the scheduler, and the agent is kept on the scheduler's resource pool as long as it maintains its connection with it.
```

Author: I have a question here. We don't really get what is a status change ? 

For me the most practical thing is to inform the scheduler each time the cpu or ram change in a certain range. Like 15% to 20% then 20% to 25%. I'm not sure it is what you meant.

We could also inform the scheduler each time a step is launched our health information. But then some steps could take a lot of time to complete and therefore begin to consume a lot of ram and cpu at any moment. 
In this case the cpu or ram usage could be 100% but the last stream told the scheduler 25%, it could be critical.

That is why we decided to stream periodically our health information. We don't get how other solution could be better.

Reviewer: A status change is an agent's state change. When the agent capacity changes in a meaningful way, or when the agent knows that it must allocate a certain capacity for handling a step or for anything else, it must notify the scheduler about it. For example, when the scheduler sends a step that consumes n cpus and M slices of memory, the agent takes that information into account and reserves that amount of resources + whatever overhead it needs to actually run the step. Then it informs the scheduler about that status change.
That's just one example, there could be other reasons for an agent capcity to drop or increase, but that's for the agent to manage.

**Launching jobs**

> An agent will try to launch every job the scheduler will give to him. Each time a job is received the agent will:

Reviewer: It should not try. Either it accepts it or it refuses it. If it accepts the job, it guarantees to the scheduler that the job will eventually run.

> - try to create an encapsulated environment from the configuration of the job

Reviewer: What is an encapsulated environment?

>     - if the configuration fail the job will considers to be failed

Reviewer: What do you mean by job configuration?

- once the configuration is done the list of command will be executed 

>     - if one command fail the job will be considered as failed

Reviewer: ```suggestion
    - if one command fails the job will be considered as failed
```

That's an implementation detail. The only thing that the scheduler cares about is if a job completed or not, and the corresponding logs.

# Summary

The agent is the powerhouse of SealCi. It receives actions and runs them in order to complete the operational part of the CI.

# References

## Actions

> A action is a defined as list of commands. actions are launched in a controlled environment, as defined by the agent configuration.

Reviewer: ```suggestion
An action is defined as a list of commands. Actions are launched in a controlled environment, as defined by the agent configuration.
```

Reviewer: The agent configuration does not defined the actions execution environment. The pipeline definition describes each action execution envronment. See https://github.com/dev-sys-do/sealci/pull/4

## Artifacts

> Artifacts is the result of the execution of a action. We consider as artifacts:
> 
> -  logs of execution of the pod
> -  any file that is created during the execution

Reviewer: ```suggestion
During their execution, actions generate artfacts, which include:

-  Execution logs
-  Files generated during the action execution
```

Reviewer: You should not introduce the Kubernetes notion if pods here.

Reviewer 2: Done

# Agent operations

## Life of an agent

> **Launching & Registering with the Scheduler**  
> Initially, the agent registers with a scheduler. After the schedulers acknowledges the registration, the agent can start processing new actions in order to accept new actions.  
> As part of the registration process, the agent and the scheduler establish a streaming, bi-directional connection.

Reviewer: ```suggestion
**Launching & Registering with the Scheduler**

Initially, the agent registers with a scheduler. As part of the registration process, the agent and the scheduler establish a streaming, bi-directional connection. 

After the schedulers acknowledges the registration, the agent is ready to accept and process new actions.

```

Reviewer 2: Done

**Health and Death**  
An agent streams health and status information to the scheduler, and the agent is kept on the scheduler's resource pool as long as it maintains its connection with it.

**Launching actions**  
Each time a action is received the agent will:

> -  create an container from the configuration of the action

Reviewer: ```suggestion
-  Create and run a container, based on the action configuration
```

Reviewer 2: Done

>    -  if the configuration fails the action will considers to be failed

Reviewer: I feel this is a bit of a tautology and I would remove that sentence.

Reviewer 2: Done

> -  once the configuration is done the list of command will be executed

Reviewer: ```suggestion
-  The list of command described in the action configuration are executed in the action container.
```

Reviewer 2: Done

>    -  for each command, an exit code will be returned to the scheduler

Reviewer: What happens when one command fails? Do we continue executing the next ones?

Reviewer 2: We don't execute the next ones. We assume that an error in one command means the failure of the action.

> -  finally, agent will be clean from any remains of the operation

Reviewer: ```suggestion
-  Once all the action commands are completed, the agent cleans the action execution environment up by deleting its container.
```

Reviewer 2: Done

> [!NOTE]  
> Each time a action stop, artifacts of what happened will be sent back. It can be:
>
> -  logs of a fail configuration setup
> -  logs of a fail execution
> -  a file created during the execution on succeed
> -  logs of a succeeded action

> 

Reviewer: You describe what artifacts are in a previous section, I'm not sure what this one brings?

Reviewer 2: Not much, except details about the different logs. I don't think this is necessary, so I've removed that part completely.

> ## Encapsulated environment

Reviewer: What is an encapsulated environment?

Reviewer 2: An encapsulated environment is the action's execution environment

# Summary
The agent is the powerhouse of SealCi. It will receive jobs and try to run them in order to complete the operational part of the CI.

# The role of the agent
The goal of this component  is to provide execution environment for the scheduler and safe and correctly configured environments.

# Agent operations

## Registering to the scheduler
The agent is a slave of the scheduler and therefore he needs to be able to communicate with it. The agent will register to the scheduler at its startup. The scheduler will then be able to send jobs to the agent.

## Health link
At any moment in its life, the agent must be able to provide information about its health status. This information will be sent to the scheduler so that it can make decisions accordingly.

> TODO: parler de la socket ouverte pour envoyer les infos de santé en continu

Author: Confirmed

## Getting jobs

> Jobs are the composition of configuration for the execution environnement and for the execution itself that will run in an encapsulated environment. The agent must have access to this environment in order to manage it.

Author: Tricky one ! When I deal with job i mean both the process and the configuration of the job written with yaml. I think it is not a problem for you if i amalgam concepts

> The scheduler will send jobs to the agent for it to execute them taking into account the health status of the agent, 

Author: I have no cat but on of my VM did explode once. Seriously, we will give chronologically health information of the agent by giving CPU and RAM information.

> TODO: changer pour dire que le scheduler ouvre une socket pour envoyer le info sur un job et qu'il y a donc une socket par job, et que si la socket est fermée le job est annulé.

Author: Not at all grpc is nice for that !

## Creating encapsulated environment
The agent will try to create an encapsulated environment from a given configuration. This mechanism is mandatory, if it fail the entire job will fail. 
If it successful, the job will process the execution configuration and run it. 

> The agent must have access to every logs of the job both from the creation of the environment and execution part. To be able to stream them to the scheduler.

Author: Yeah it as been modified but in the end we want the logs of jobs to appear somewhere. In fact we aren't going to stream them as they come. We are going to send all logs that happened once the job is done. I think the scheduler will also give jobs' logs back to the controller. Not sure about that we can discuss it if you want !

## Running jobs

> Once the environment has been created and configured and the job has been received, the agent must execute the job.

Reviewer 3: [scheduler team]
Does the received job contain the environment ocnfiguration ?

My first thought is that there is a default configuration, and the Job may contain overloading environment configuration data such as (but not limited to?) environment variables ?

So the environment could be configured through scripts ?

Author: As we go in the project we think we will use containers to manage those processes. So the configuration should be the image the container will use.

TODO: maybe mouve the part about the introduction of logs streaming from the creating env here

## Streaming informations and logs

> At each stage of the job's progress, the agent must be able to stream information to the scheduler thanks to an open connection throughout the life of the agent.

Author: You are right. We should let the Operator of the CI chose a timeout for jobs. Thanks !
